clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4486,11 +4489,66 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   if (IsIAMCU && types::isCXX(Input.getType()))
     D.Diag(diag::err_drv_clang_unsupported) << "C++ for IAMCU";
 
+  // Handle depscan.
+  if (Job.getKind() == Action::DepscanJobClass) {
+    CmdArgs.push_back("-cc1depscan");
+
+    // Pass depscan related options to cc1depscan.
+    const OptSpecifier DepScanOpts[] = {
+        options::OPT_fdepscan_EQ,
+        options::OPT_fdepscan_share_EQ,
+        options::OPT_fdepscan_share_parent,
+        options::OPT_fdepscan_share_parent_EQ,
+        options::OPT_fno_depscan_share,
+        options::OPT_fdepscan_share_stop_EQ,
+        options::OPT_fdepscan_daemon_EQ,
+        options::OPT_fdepscan_prefix_map_EQ,
+        options::OPT_fdepscan_prefix_map_sdk_EQ,
+        options::OPT_fdepscan_prefix_map_toolchain_EQ};
+
+    Args.AddAllArgs(CmdArgs, DepScanOpts);
+
+    assert(Output.isFilename() && "Depscan needs to have file output");
+    CmdArgs.push_back("-o");
+    CmdArgs.push_back(Output.getFilename());
+    CmdArgs.push_back("-cc1-args");
+  }
+
   // Invoke ourselves in -cc1 mode.
   //
   // FIXME: Implement custom jobs for internal actions.
   CmdArgs.push_back("-cc1");
 
+  // Handle full response file input.
+  if (Inputs.front().getType() == types::TY_ResponseFile) {
+    // Render response file input first.
+    assert(Inputs.size() == 1 && "Only one response file input");
+    auto &II = Inputs.front();
+    assert(II.isFilename() && "Should be response file");
+    SmallString<128> InputName("@");
+    InputName += II.getFilename();
+    CmdArgs.push_back(C.getArgs().MakeArgString(InputName));
+
+    // FIXME: Overwrite the output filename. To fix this, we need to somehow
+    // route the final output type into Depscan ConstructJob.
+    if (Output.isFilename()) {
+      CmdArgs.push_back("-o");
+      CmdArgs.push_back(Output.getFilename());
+    }
```

Comment:
Can this be removed now, since the `-cc1depscan` is getting the `-o`?

Reply:
`-cc1depscan` doesn't get the correct output file for cc1 argument so this `-o` is passed after response file with the correct output filename. `FIXME` explained how to fix it. It is not hard but need to thread another argument through lots of functions. This is a small overwrite so I think it should be fine for now.

Reply:
Okay; the FIXME wasn't clear to me since it doesn't describe the problem. I assumed this was left over from before you'd added the 'combine' step.

I suggest something like this:
```
// FIXME: The -cc1depscan job should include the -o for the -cc1 job that follows
// it in -cc1-args, but it doesn't currently have access to that filename. To fix this,
// we need to somehow route the final output type into Depscan ConstructJob.
```